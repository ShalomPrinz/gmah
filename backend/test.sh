#!/bin/bash

if [[ $1 == "--coverage" ]]; then
    coverage run -m unittest discover tests
    coverage html
else
    python -m unittest discover tests
fi
