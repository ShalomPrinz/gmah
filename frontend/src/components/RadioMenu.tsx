import { useState } from "react";
import FormCheck from "react-bootstrap/FormCheck";

import { ConditionalList } from ".";

interface RadioButton {
  id: string;
  text: string;
  value: string;
}

interface RadioMenuProps {
  buttons: RadioButton[];
  menuId: string;
  onSelect: (value: string) => void;
}

function RadioMenu({ buttons, menuId, onSelect }: RadioMenuProps) {
  const initialSelected = buttons?.length ? buttons[0].value : "";
  const [selected, setSelected] = useState(initialSelected);

  const onButtonSelect = (value: string) => {
    setSelected(value);
    onSelect(value);
  };

  const buttonCallback = ({ id, text, value }: RadioButton) => (
    <FormCheck
      id={id}
      className="fs-5 my-0"
      checked={selected === value}
      label={text}
      name={menuId}
      onChange={(e) => onButtonSelect(e.target.value)}
      reverse
      type="radio"
      value={value}
    />
  );

  return (
    <form>
      <ConditionalList itemCallback={buttonCallback} list={buttons} />
    </form>
  );
}

export default RadioMenu;
