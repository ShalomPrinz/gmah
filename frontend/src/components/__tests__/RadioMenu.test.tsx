import { describe, expect, it, vi } from "vitest";

import { render, screen, setupUser } from "../../../test";

import RadioMenu from "../RadioMenu";

describe("RadioMenu", () => {
  describe("snapshot", () => {
    it("should render empty radio menu", () => {
      const { asFragment } = render(
        <RadioMenu buttons={[]} menuId="" onSelect={() => {}} />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render one radio button", () => {
      const { asFragment } = render(
        <RadioMenu
          buttons={[{ id: "1", text: "button text", value: "hello world" }]}
          menuId="menu id"
          onSelect={() => {}}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render three radio buttons", () => {
      const { asFragment } = render(
        <RadioMenu
          buttons={[
            { id: "1", text: "button text", value: "hello world" },
            { id: "2", text: "other text", value: "select me" },
            { id: "3", text: "another", value: "option 3" },
          ]}
          menuId="menu id"
          onSelect={() => {}}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("behavior", () => {
    it("should select second option and get its value", async () => {
      const user = setupUser();
      const secondButtonValue = "to be selected";
      const onSelect = vi.fn((value: string) => {});
      render(
        <RadioMenu
          buttons={[
            { id: "1", text: "button text", value: "hello world" },
            { id: "2", text: "other text", value: secondButtonValue },
          ]}
          menuId="menu id"
          onSelect={onSelect}
        />
      );

      const element = screen.getAllByRole("radio")[1];
      await user.click(element);

      expect(onSelect).toBeCalledWith(secondButtonValue);
    });

    it("should select second option and then third option", async () => {
      const user = setupUser();
      const thirdButtonValue = "to be selected";
      const onSelect = vi.fn((value: string) => {});
      render(
        <RadioMenu
          buttons={[
            { id: "1", text: "button text", value: "hello world" },
            { id: "2", text: "nothing here", value: "irrelevant" },
            { id: "3", text: "other text", value: thirdButtonValue },
          ]}
          menuId="menu id"
          onSelect={onSelect}
        />
      );

      const buttons = screen.getAllByRole("radio");
      await user.click(buttons[1]);
      await user.click(buttons[2]);

      expect(onSelect).toBeCalledWith(thirdButtonValue);
    });
  });
});
