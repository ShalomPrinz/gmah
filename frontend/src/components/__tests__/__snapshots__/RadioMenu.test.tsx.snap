// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RadioMenu > snapshot > should render empty radio menu 1`] = `
<DocumentFragment>
  <form />
</DocumentFragment>
`;

exports[`RadioMenu > snapshot > should render one radio button 1`] = `
<DocumentFragment>
  <form>
    <div
      class="fs-5 my-0 form-check form-check-reverse"
    >
      <input
        checked=""
        class="form-check-input"
        id="1"
        name="menu id"
        type="radio"
        value="hello world"
      />
      <label
        class="form-check-label"
        for="1"
        title=""
      >
        button text
      </label>
    </div>
  </form>
</DocumentFragment>
`;

exports[`RadioMenu > snapshot > should render three radio buttons 1`] = `
<DocumentFragment>
  <form>
    <div
      class="fs-5 my-0 form-check form-check-reverse"
    >
      <input
        checked=""
        class="form-check-input"
        id="1"
        name="menu id"
        type="radio"
        value="hello world"
      />
      <label
        class="form-check-label"
        for="1"
        title=""
      >
        button text
      </label>
    </div>
    <div
      class="fs-5 my-0 form-check form-check-reverse"
    >
      <input
        class="form-check-input"
        id="2"
        name="menu id"
        type="radio"
        value="select me"
      />
      <label
        class="form-check-label"
        for="2"
        title=""
      >
        other text
      </label>
    </div>
    <div
      class="fs-5 my-0 form-check form-check-reverse"
    >
      <input
        class="form-check-input"
        id="3"
        name="menu id"
        type="radio"
        value="option 3"
      />
      <label
        class="form-check-label"
        for="3"
        title=""
      >
        another
      </label>
    </div>
  </form>
</DocumentFragment>
`;
